### PR TITLE
[core] Pmd7 test cleanups

### PR DESCRIPTION
**Pull request by oowekyala**
## Describe the PR

Cleanup tests by introducing new utilities, like `BaseParsingHelper#executeRule(Rule,String) -> Report`. This removes a lot of usage sites of RuleContext, and prepares further refactoring of this object.

This also makes many places in the violation reporting codebase more defensive in requiring their parameters be non-null. Many tests build dysfunctional violations by passing a null node or null file name or whatnot, so these needed to be cleaned up.

ParametricRuleViolation is changed to not require a rule context, but instead a file name. This removes a lot of places where we create a rule context just to set the filename or the language version. The RuleViolationFactory interface is also changed, to only require an implementation of createViolation and suppressOrNull.

The longer-term idea, is to expunge RuleContext of all state, by moving getSourceCodeFile and getLanguageVersion onto the Node interface. Given that, RuleContext doesn't need to store this state, and so has no purpose anymore. I think, we can repurpose it to be the API for rules to report violations. It would basically have the overloads of `addViolation` that AbstractRule has for now, but clearer. In the end, the idea is for rules to have type-safe access to a RuleContext and report violations directly on it instead of using `AbstractRule#addViolation(Object, ...)`.

## Related issues

- Fixes #

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by travis)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">No java rules are changed!</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by oowekyala** [comment]
@adangel I'm going to remove the API-breaking changes from RuleTst (so that we can decide on that later) and merge this PR this week

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/ParametricRuleViolation.java [hunk 1]
```diff
@@ -34,26 +35,39 @@
     // RuleViolationFactory to support identifying without a Node, and update
     // Rule base classes too.
     // TODO we never need a node. We just have to have a "position", ie line/column, or offset, + file, whatever
+
+    /**
+     * @deprecated Use {@link #ParametricRuleViolation(Rule, String, Node, String)}
+     */
+    @Deprecated
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/ParametricRuleViolation.java:42]: Deprecated or DeprecatedUntil700?
If deprecated, we need to merge this back to master -> API change...

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/ParserOptions.java [hunk 2]
```diff
@@ -30,11 +37,11 @@ public boolean equals(Object obj) {
...
         }
         final ParserOptions that = (ParserOptions) obj;
-        return this.suppressMarker.equals(that.suppressMarker);
+        return this.getSuppressMarker().equals(that.getSuppressMarker());
     }
 
     @Override
     public int hashCode() {
-        return suppressMarker != null ? suppressMarker.hashCode() : 0;
+        return getSuppressMarker() != null ? getSuppressMarker().hashCode() : 0;
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/lang/ParserOptions.java:45]: I think, we don't need a null check here anymore, do we?

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/RuleViolationFactory.java [hunk 3]
```diff
@@ -34,8 +44,43 @@
...
+        addViolation(ruleContext, rule, node, message, node.getBeginLine(), node.getEndLine(), args);
+    }
+
+
+    @Deprecated
+    default void addViolation(RuleContext ruleContext, Rule rule, @NonNull Node node, @NonNull String message, int beginLine, int endLine, @NonNull Object[] args) {
+        Objects.requireNonNull(node);
+        Objects.requireNonNull(message);
+        Objects.requireNonNull(args);
+        RuleViolation rv = formatViolation(rule, node, ruleContext.getSourceCodeFilename(), message, args);
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/RuleViolationFactory.java:58]: `ruleContext.getSourceCodeFilename()` only returns the simple name. We need to use here `String.valueOf(ruleContext.getSourceCodeFile())`. I'll fix that on pmd/7.0.x and add a unit test (hopefully not forgetting that there is also windows...)

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/RuleViolationFactory.java [hunk 4]
```diff
@@ -34,8 +44,43 @@
      * @param args
      *            arguments to embed in the rule violation message
      */
-    void addViolation(RuleContext ruleContext, Rule rule, @Nullable Node node, String message, Object[] args);
+    @Deprecated
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/RuleViolationFactory.java:47]: Probably DeprecatedUntil700 as well?

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/ParserOptions.java [hunk 5]
```diff
@@ -4,20 +4,27 @@
...
+
 /**
  * Represents a set of configuration options for a {@link Parser}. For each
  * unique combination of ParserOptions a Parser will be used to create an AST.
  * Therefore, implementations must implement {@link Object#equals(Object)} and
  * {@link Object#hashCode()}.
  */
 public class ParserOptions {
-    protected String suppressMarker;
+    private String suppressMarker = PMD.SUPPRESS_MARKER;
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/lang/ParserOptions.java:20]: I think, we should deprecate this field on master...

**Pull request merged**